src/lib.rs

```diff
@@ -333,35 +365,58 @@ impl Macaroon {
         );
     }
 
-    /// Serialize the macaroon using the serialization format provided
-    pub fn serialize(&self, format: serialization::Format) -> Result<Vec<u8>> {
+    /// Serialize the macaroon using the serialization [Format] provided
+    ///
+    /// For V1 and V2, the binary format will be encoded as URL-safe base64 with padding
+    /// (`base64::URL_SAFE`). For V2JSON, the output will be JSON.
+    pub fn serialize(&self, format: serialization::Format) -> Result<String> {
         match format {
             serialization::Format::V1 => serialization::v1::serialize(self),
             serialization::Format::V2 => serialization::v2::serialize(self),
             serialization::Format::V2JSON => serialization::v2json::serialize(self),
         }
     }
 
-    /// Deserialize a macaroon
-    pub fn deserialize(data: &[u8]) -> Result<Macaroon> {
-        if data.is_empty() {
+    /// Deserialize an encoded macaroon token, inferring the [Format]
+    ///
+    /// For V1 and V2 tokens, this assumes base64 encoding, in either "standard" or URL-safe
+    /// encoding, with or without padding.
+    pub fn deserialize<T: AsRef<[u8]>>(token: T) -> Result<Macaroon> {
+        if token.as_ref().is_empty() {
             return Err(MacaroonError::DeserializationError(
                 "empty token provided".to_string(),
             ));
         }
-        let macaroon: Macaroon = match data[0] as char {
-            '{' => serialization::v2json::deserialize(data)?,
-            '\x02' => serialization::v2::deserialize(data)?,
-            'a'..='z' | 'A'..='Z' | '0'..='9' | '+' | '-' | '/' | '_' => {
-                serialization::v1::deserialize(data)?
+        let mac: Macaroon = match token.as_ref()[0] as char {
+            '{' => serialization::v2json::deserialize(token.as_ref())?,
+            _ => {
```

Comment:
Should we allow for the possibility of leading whitespace and strip that, or just specify that such a token will be rejected?

We'd have that case drop to the default binary one below which is probably not right.

Reply:
I'm in favor of keeping the existing behavior, at least for now, which matches libmacaroon. I think we should try to keep this logic as simple as possible; to make V2JSON parsing more flexible we could add a format-specific deserialization method in the future.

I updated the doc string on this method, and added usage examples (which also form a doctest).